python_modules/dagster-graphql/dagster_graphql/implementation/utils.py

```diff
@@ -133,12 +127,32 @@ def has_permission_for_asset_graph(
 
     if not location_names:
         return context.has_permission(permission)
-    else:
+
+    # if we have permission for all locations relevant to the asset graph, we're good
+    if all(
+        context.has_permission_for_location(permission, location_name)
+        for location_name in location_names
+    ):
+        return True
+
+    # No need to check individual asset keys if we don't have owner permissions
+    if not context.viewer_has_any_owner_definition_permissions(permission):
+        return False
+
+    if asset_keys:
+        # if we have owner permissions and are checking specific asset keys, check that we own all of
+        # the asset keys
         return all(
-            context.has_permission_for_location(permission, location_name)
-            for location_name in location_names
+            context.has_permission_for_definition(permission, asset_graph.get(key))
+            for key in asset_keys
+            if asset_graph.has(key)
         )
 
+    # enumerate all the asset nodes and check permissions
+    return all(
+        context.has_permission_for_definition(permission, node) for node in asset_graph.asset_nodes
+    )
```

Comment:
hm this seems potentially very expensive

Are there actually any callsites that don't pass in a list of asset keys here? Can we make that argument required?